### ISSUE TITLE
Unit testing with takeLatest or takeEvery

### ISSUE THREAD

**Issue by bpmckee** [yelouafi/redux-saga#318]
From the examples I've seen, if you use a watcher saga that uses `takeLatest` or `takeEvery`, the tests do not ensure the saga is run after seeing the correct action dispatched.

For example here are my sagas.
```js
function* getDataSaga(action) {
  try {
    const response = yield call(api.getData, action.payload);
    yield put(getDataSucceeded(response));
  } catch (e) {
    yield put(getDataFailed(e));
  }
}

function* watchGetDataSaga() {
  yield* takeLatest(GET_DATA, getDataSaga);
}
```

I could write some tests like this:
```js
describe('My Saga', () => {
  it('should first call the getData api', () => {
    // arrange
    const action = {
      type: GET_DATA,
      payload: { foo: 'bar' },
    };
    const iterator = getDataSaga(action);
    const expectedYield = call(api.getData, action.payload);

    // act
    const actualYield = iterator.next().value;

    // assert
    expect(actualYield).to.deep.equal(expectedYield);
  });
});
```

However, I can't write tests that make sure the saga is kicked off from a GET_DATA action.  It seems to me the only way to do this sort of test is to use a `while(true)` in my `getDataSaga` instead of having the `watchGetDataSaga`.  Is this the case, or is there a way around this?

**Comment by yelouafi** [comment]
You can treat takeEvery/takeLatest either as a single call; in this case you can use it with `yield call(takeEvery, ACTION, saga)` or (`yield fork(takeEevry,..)` and the test could be just as a normal call (or fork) Effect test.

If you want to test the internals of the Generator (when you use `yield*`) you can find the implementation here http://yelouafi.github.io/redux-saga/docs/advanced/Concurrency.html

**Comment by bpmckee** [comment]
Ok, so I'm still having trouble.  Assuming my same saga definitions from above, here's the test I have that's failing.

```js
it('should trigger on GET_DATA', () => {
  // arrange
  const iterator = watchGetDataSaga();
  const expectedYield = call(takeEvery, GET_DATA, getDataSaga);

  // act
  const actualYield = iterator.next().value;

  // assert
  expect(actualYield).to.deep.equal(expectedYield);
});
```

This threw some nasty errors
```
AssertionError: expected { Object (@@redux-saga/IO, TAKE) } to deeply equal { Object (@@redux-saga/IO, CALL) }
```
```
{
  -  "TAKE": {
  -    "channel": [null]
  -    "pattern": "GET_DATA"
  +  "CALL": {
  +    "args": [
  +      "GET_DATA"
  +      {}
  +    ]
  +    "context": [null]
  +    "fn": [Function]
     }
}
```

**Comment by npbee** [comment]
I think what @yelouafi is saying is that instead of 

```
yield* takeEvery(...)
```

...you can do

```
yield call(takeEvery, ...)
```
So if you're testing that the saga yields a `call`, you'll need to update your `watchGetDataSaga` to actually yield a `call` effect.  Here's a fiddle:  https://jsfiddle.net/npbee/jksq12gg/1/.

Also, in your snippet above you're testing a call of `takeEvery` when your code has `takeLatest`.  It's not the root of the problem but might be causing more confusion for you.

**Comment by bpmckee** [comment]
Hey that worked!  Thank you guys.  I'll close this issue now since the question is resolved.

**Comment by metrue** [comment]
@npbee 

quick question here, mybe this is dumb one.

but what's the deference between 

```
yield* takeEvery
```

and 

```
yield call(takeEvery, ...)
```

do I need always use <code> call </code> to make testing work ?

**Comment by npbee** [comment]
Not dumb!  I'm no expert myself, but I'll explain as I understand it.  

Both of these basically have the same functional outcome.  You can test both in the way we discussed above but the return values are different, so the tests would change.  

The `yield*` expression is just a plain JS [feature](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/yield*), and it essentially just allows to you to `yield` to nested iterables.  In this context, doing `yield* takeEvery` would mean that you're yielding to the every value that the `takeEvery` iterable _itself_ yields.  So to test `yield* takeEvery`, you'd be testing for what values `takeEvery` itself yields back to you, which means you need to know the internals of `takeEvery`.  

When using `yield call(takeEvery, ..)`, that's a declarative effect from redux-saga and it will only ever `yield` one value and that is a `CALL` effect.  You can easily test that the correct effect was yielded.  

So long story short:  functionally they are basically the same but testing `call` effects are easier.

I made another fiddle that might be more clear:  https://jsfiddle.net/npbee/Lqreq12b/2/.  

Hope that helps!

**Comment by metrue** [comment]
@npbee 

what a great explanation here, I totally understand it now.  and thanks for you great example.

**Comment by hwangar** [comment]
Not sure if this is working... yeah, the test is passing but code does not work:

`
const { createStore, applyMiddleware } = require('redux');
const { default: createSagaMiddleware } = require('redux-saga');
const { takeEvery, call } = require('redux-saga/effects');

const sagaMiddleware = createSagaMiddleware();
const store = createStore((state = {}) => state, undefined, applyMiddleware(sagaMiddleware));

const helloWorld = function() {
    console.log('Hello world');
}
const watchHello = function* () {
    //yield takeEvery('hello', helloWorld);
    yield call(takeEvery, 'hello', helloWorld);
}

sagaMiddleware.run(watchHello);

store.dispatch({ type: 'hello' });
`

Commented code will show greeting, but uncommented does not :(

**Comment by hwangar** [comment]
Figured out myself :)

The correct/current way should be:

```
yield fork(takeEvery, 'hello', helloWorld);
```

So testing with just:

```
assert.deepEqual(
    gen.next().value, 
    fork(takeEvery, 'hello', helloWorld)
);
```

**Comment by Andarist** [comment]
Actually preferred way is to yield takeEvery. Are u on the latest version of redux-saga? How do u import takeEvery?

**Comment by hwangar** [comment]
See previous comment: `const { takeEvery, call } = require('redux-saga/effects');`

Yeah, I know, this way is intended to test watchers sagas

**Comment by Andarist** [comment]
You have mentioned that this is working for u
```js
const { takeEvery } = require('redux-saga/effects');
// ...
yield fork(takeEvery, 'hello', helloWorld);
```
and it shouldnt I think.

Correct working way is to `yield takeEvery('hello', helloWorld);`

**Comment by hwangar** [comment]
Not sure Andarist, TBH my aim was to write a test for my first comment's code following the indications of this thread... (there's even a gist out there), but instead of launching takeEvery with call I found it works with fork:

* Saga is performed on action as intended
* Test works good deep equaling

**Comment by Andarist** [comment]
Could u share a reproduced code on something like https://codesandbox.io ? `takeEvery` effect is [already wrapped in a `fork` for you under the hood](https://github.com/redux-saga/redux-saga/blob/02bda102e299f882a09db06947cd0a2930951157/src/internal/io.js#L184-L186).

So according to what u have written it becomes something like `yield fork(fork(takeEveryHelper, 'hello', helloWorld))` and this certainly shouldnt work. 

There is a possibility though that you import `takeEvery` from the `redux-saga` and not `redux-saga/effects`, but that way is deprecated for quite some time.

**Comment by hwangar** [comment]
Of course man!

https://codesandbox.io/s/vlvMJqJ5

I don't think it's a bad import, I checked twice

**Comment by Andarist** [comment]
@hwangar 
Thank you for the the raising this up. You have unveiled a really subtle bug, need to figure out how to fix it properly now :).

As to ur origina issue though - preferred way is to:
```js
yield takeEvery("task", task);
// ...
assert.deepEqual(watchTask().next().value, takeEvery("task", task));
```

**Comment by hwangar** [comment]
It works like a charm, thank you @Andarist

**Comment by kencckw** [comment]
I got Typescript error when I `call(takeEvery, ...)`

The only way I can do is  `takeEvery:any`

Please help.

`Argument of type '{ <A extends Action>(pattern: Pattern, worker: HelperFunc0<A>): ForkEffect; <A, T1>(pattern: Patt...' is not assignable to parameter of type 'CallEffectNamedFn<{ [x: string]: Func2<string, (action: any) => IterableIterator<any>>; }, string>'.
  Type '{ <A extends Action>(pattern: Pattern, worker: HelperFunc0<A>): ForkEffect; <A, T1>(pattern: Patt...' is not assignable to type '{ context: { [x: string]: Func2<string, (action: any) => IterableIterator<any>>; }; fn: string; }'.
    Property 'context' is missing in type '{ <A extends Action>(pattern: Pattern, worker: HelperFunc0<A>): ForkEffect; <A, T1>(pattern: Patt...'.`

**Comment by bstro** [comment]
I have the same issue as @kencckw

**Comment by Andarist** [comment]
`call(takeEvery, ...)` is old recommendation, please just use `takeEvery(...)` from the `redux-saga/effects` 'subpackage'
